lib/jasmine/jasmine.ts

```diff
@@ -1,32 +1,97 @@
-// synchronously means that the current zone does not get cleared. This
-// results in a chain of nested zones, which makes it hard to reason about
-// it. We override the `clearStack` method which forces jasmine to always
-// drain the stack before next test gets executed.
-(<any>jasmine).QueueRunner = (function (SuperQueueRunner) {
-  const originalZone = Zone.current;
-  // Subclass the `QueueRunner` and override the `clearStack` method.
+(() => {
+  // Patch jasmine's it and fit functions so that the `done` wrapCallback always resets the zone
+  // to the jasmine zone, which should be the root zone. (angular/zone.js#91)
+  if (!Zone) throw new Error("Missing: zone.js");
+  const SyncTestZoneSpec: {new (name: string): ZoneSpec} = Zone['SyncTestZoneSpec'];
+  if (!SyncTestZoneSpec) throw new Error("Missing: SyncTestZoneSpec");
+  const ProxyZoneSpec: {new (): ZoneSpec} = Zone['ProxyZoneSpec'];
+  if (!ProxyZoneSpec) throw new Error("Missing: ProxyZoneSpec");
+  if (!jasmine) throw new Error("Missing: jasmine.js");  
+  if (jasmine['__zone_patch__']) throw new Error("'jasmine' has already been patched with 'Zone'.");
+  jasmine['__zone_patch__'] = true;
 
-  function alwaysClearStack(fn) {
-    const zone: Zone = Zone.current.getZoneWith('JasmineClearStackZone')
-        || Zone.current.getZoneWith('ProxyZoneSpec')
-        || originalZone;
-    zone.scheduleMicroTask('jasmineCleanStack', fn);
+  const ambientZone = Zone.current;
+  // Sync zone will throw on any async operations. 
+  // Will be used when executing contents of describe.
+  const syncZone = ambientZone.fork(new SyncTestZoneSpec('jasmine.describe'));
+
+  // This is the zone which will be used for running individual tests.
+  // It will be a proxy zone, so that the tests function can retroactivly install
+  // different zones. 
+  let testProxyZone: Zone = null;
+
+  // Monkey patch all of the jasmine DSL so that each function runs in appropriate zone.
+  const jasmineEnv = jasmine.getEnv();
+  ['desribe', 'xdescribe', 'fdescribe'].forEach((methodName) => {
+    var delegate: Function = jasmineEnv[methodName];
+    jasmineEnv[methodName] = function(description: string, specDefinitions: Function) {
+      return delegate.call(this, description,  wrapDescribeInZone(specDefinitions));
+    } 
+  });
+  ['it', 'xit', 'fit'].forEach((methodName) => {
+    var delegate: Function = jasmineEnv[methodName];
+    jasmineEnv[methodName] = function(description: string, specDefinitions: Function) {
+      return delegate.call(this, description, wrapTestInZone(specDefinitions));
+    } 
+  });
+  ['beforeEach', 'afterEach'].forEach((methodName) => {
+    var delegate: Function = jasmineEnv[methodName];
+    jasmineEnv[methodName] = function(specDefinitions: Function) {
+      return delegate.call(this, wrapTestInZone(specDefinitions));
+    } 
+  });
+
+  function wrapDescribeInZone(fn: Function): Function {
+    // Used for executing the describe content. No async work is allowed here.
```

Comment:
Should this comment (and the one for `wrapInTestZone`) be function-description comments? I.e., 
```ts
/** Gets a function wrapping the body of a Jasmine `describe` block to execute in a synchronous-only zone. */
function wrapDescribeInZone(fn: Function) {
  ...
}
```